File: Medium/Javascript/988. Smallest String Starting From Leaf.js
/**You are given the root of a binary tree where each node has a value in the range [0, 25] representing the letters 'a' to 'z'.

Return the lexicographically smallest string that starts at a leaf of this tree and ends at the root.

As a reminder, any shorter prefix of a string is lexicographically smaller.

For example, "ab" is lexicographically smaller than "aba".
A leaf of a node is a node that has no children. */

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string}
 */
function numToal(number) {
  return String.fromCharCode(97 + number);
}

function smallestFromLeaf(root) {
  const queue = [];
  queue.push([root, numToal(root.val)]);
  const results = [];

  while (queue.length > 0) {
    const [cur, path] = queue.shift();
    if (!cur.left && !cur.right) {
      results.push(path.split("").reverse().join(""));
    }
    if (cur.left) {
      queue.push([cur.left, path + numToal(cur.left.val)]);
    }
    if (cur.right) {
      queue.push([cur.right, path + numToal(cur.right.val)]);
    }
  }
  results.sort();
  return results[0];
}

File: Medium/python/988. Smallest String Starting From Leaf.py
"""
You are given the root of a binary tree where each node has a value in the range [0, 25] representing the letters 'a' to 'z'.

Return the lexicographically smallest string that starts at a leaf of this tree and ends at the root.

As a reminder, any shorter prefix of a string is lexicographically smaller.

For example, "ab" is lexicographically smaller than "aba".
A leaf of a node is a node that has no children.
"""

import collections

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def numToal(self, number):
        return chr(97 + number)

    def smallestFromLeaf(self, root: TreeNode) -> str:
        queue = collections.deque()
        queue.append((root, self.numToal(root.val)))
        results = []
        
        while queue:
            cur, path = queue.popleft()
            if not cur.left and not cur.right:
                results.append(path[::-1])
            if cur.left:
                queue.append((cur.left, path + self.numToal(cur.left.val)))
            if cur.right:
                queue.append((cur.right, path + self.numToal(cur.right.val)))
        results.sort()
        return results[0]
